### PR TITLE
Musique module

### LINKED ISSUE #5: [1.0] MusicAlbum - setup class
Create a MusicAlbum class in a separate .rb file.
All MusicAlbum class properties visible in the diagram should be defined and set up in the constructor method.

### LINKED ISSUE #8: [1.0] Genre - setup class
Create a Genre class with an association to the Item class (in a separate .rb file).
All Genre class properties visible in the diagram should be defined and set up in the constructor method.

### LINKED ISSUE #11: [1.0] MusicAlbum - implement methods
Implement the following method:
can_be_archived?() in the MusicAlbum class
should override the method from the parent class
should return true if parent's method returns true AND if on_spotify equals true
otherwise, it should return false

### LINKED ISSUE #14: [1.0] Genre - implement methods
Implement method:
add_item method in the Genre class
should take an instance of the Item class as an input
should add the input item to the collection of items
should add self as a property of the item object (by using the correct setter from the item object)

### LINKED ISSUE #20: [1.0] Genre - Add unit tests
Add unit tests for all implemented methods.

### LINKED ISSUE #21: [1.0] MusicAlbum - Add unit tests
Add unit tests for all implemented methods.

### LINKED ISSUE #23: [1.0] MusicAlbum - List all/add MusicAlbums
Add methods to make these options available:
List all MusicAlbums.
Add a MusicAlbum.

### PR DESCRIPTION
**Pull request by masangana**
## Todo List

- Create MusicAlbum class 
- Create Genre class
- All MusicAlbum class properties visible in the diagram should be defined
- set up in the constructor method
- All Genre class properties visible in the diagram should be defined
- set up in the constructor method.
- Implement methods:
add_item method in the Genre class
can_be_archived?() in the MusicAlbum class

- Add unit tests for all implemented methods.

### CODE REVIEW

**Comment by Gopxfs** [comment]
## Hi @masangana  :wave:

Great job so far!

## Highlights :rocket:
:heavy_check_mark: Created MusicAlbum class.
:heavy_check_mark: Created Genre class.
:heavy_check_mark: Well done adding the properties to MusicAlbum & Genre!


# Status: REQUIRED CHANGES 🟡 🟡 🟡 
Your project is looking good, but not ready yet. Please implement the following changes so you can merge them:
1. To follow the project requirements, please add unit tests for all your class methods. For Genre, please test #add_item, and for MusicAlbum, please test #can_be_archived?.
2. The method #can_be_archived? inside 'item.rb' needs to be private, as specified on the UML class diagram, as well as the one who inherits from it in MusicAlbum.
3. Kindly adjust the Item class properties for 1-to-many relationships, as they shouldn't be set in the constructor method. Instead, they should have a custom setter method created.

> Please check below for more specific information.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@Gopxfs` if something is not 100% clear.

File: item.rb [hunk 1]
```diff
@@ -21,18 +24,13 @@ def add_label(label)
...
-  private
-
-  attr_reader :id
-  attr_accessor :archived
-
   def can_be_archived?
     return true if DateTime.now.year - @publish_date.year > 10
 
     false
   end
```
Comment (Gopxfs) [inline item.rb]: 2. The method #can_be_archived? inside 'item.rb' needs to be private, as specified on the UML class diagram below.
image
As well as the MusicAlbum's:
image

File: item.rb [hunk 2]
```diff
@@ -3,10 +3,13 @@
...
+  def initialize(id = Random.rand(1..1000))
     @id = id
-    @publish_date = Date.parse(publish_date) # publish_date should be a string = 'DD-MM-YYYY' or 'DD/MM/YYYY'
-    @archived = archived
+    @genre = nil
+    @author = nil
+    @label = nil
+    @publish_date = nil
+    @archived = false
   end
```
Comment (Gopxfs) [inline item.rb]: image
3. Please check the exception, these properties (genre, author, and label) shouldn't be on the constructor. Kindly use the setter methods.

File: item.rb (hunk 1 above)
Comment (masangana) [reply item.rb]: Updated and test validated for this method

File: item.rb [hunk 3]
```diff
@@ -1,27 +1,36 @@
...
+    )
   end
 
   def add_label(label)
-    @label = label
+    label.is_a?(Label) && @label.nil? && (
+      @label = label
+      label.add_item(self)
+    )
   end
```
Comment (masangana) [inline item.rb:32]: Create customer methods

File: item.rb (hunk 2 above)
Comment (masangana) [reply item.rb]: Updated. 
And created custom setter methods for them. [link ](https://github.com/Gopxfs/catalog-of-my-things/pull/51/commits/97e356aa0f76d90a4a56b593761f6db765a07536#r963509766)

**Comment by Gopxfs** [comment]
## Hi @masangana  :wave:

We are almost there!

## Highlights :rocket:
:heavy_check_mark: Implemented tests to all methods.
:heavy_check_mark: Great addition to the setter methods on Item class.

# Status: REQUIRED CHANGES 🟡 🟡 🟡 
Your project is looking good, but not ready yet. Please implement the following changes so you can merge them:
1. The method #can_be_archived? inside 'item.rb' needs to be private, as specified on the UML class diagram, as well as the one who inherits from it in MusicAlbum.
2. Kindly add the @publish_date to the constructor on the Item class, it shouldn't be initialized as nil.

> Please check below for more specific information.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@Gopxfs` if something is not 100% clear.

File: item.rb (hunk 3 above)
Comment (Gopxfs) [reply item.rb:32]: Nice implementation @masangana

File: item.rb (hunk 1 above)
Comment (Gopxfs) [reply item.rb]: > 2. The method #can_be_archived? inside 'item.rb' needs to be private, as specified on the UML class diagram below.
>    image
>    As well as the MusicAlbum's:
>    image

Please add the keyword 'private' and put all private methods below it. It should look like:

private

def can_be_archived?
   ... [logic] ...
end

File: item.rb [hunk 4]
```diff
@@ -1,38 +1,42 @@
...
 class Item
-  attr_accessor :publish_date
+  attr_reader :id, :genre, :publish_date, :archived, :author, :label
 
-  def initialize(id, publish_date, archived: false)
+  def initialize(id = Random.rand(1..10_000))
     @id = id
-    @publish_date = Date.parse(publish_date) # publish_date should be a string = 'DD-MM-YYYY' or 'DD/MM/YYYY'
-    @archived = archived
+    @publish_date = nil
```
Comment (Gopxfs) [inline item.rb]: @publish_date shouldn't be initialized as nil, please add it to the constructor

**Comment by Gopxfs** [comment]
Approved

**Pull request merged**